### PR TITLE
machinectl compliant NixOS installation

### PR DESCRIPTION
**Pull request by ck3d**
* Patched _nixos-install_ to get a system-nspawn complient NixOS installation.
* Patched NixOS option _boot.isContainer_ to cover the incompatibility of resolvconf and networkd.
* Patched NixOS activation script "var" to deactivate immutability of /var/empty in case of a container

###### Motivation for this change
Simple installation of NixOS into a systemd controlled container via machinectl.
Used networkd to get working network setup without additional configuration.

see also #9884 and #35364

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers
cc @nbarbey 
cc @

### CODE REVIEW

**Comment by markuskowa** [comment]
@GrahamcOfBorg test systemd-machinectl

File: nixos/doc/manual/installation/installing-nspawn-container.xml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+<section xmlns="http://docbook.org/ns/docbook"
+         version="5.0"
+         xml:id="sec-installing-nspawn-container">
+ <title>Installing into a nspawn container</title>
+
+ <para>
+  Installing NixOS into a systemd nspawn container possible if nix is installed
```
Comment (markuskowa) [inline nixos/doc/manual/installation/installing-nspawn-container.xml]:
```suggestion
  Installing NixOS into a systemd nspawn container is possible if nix is installed
```

File: nixos/doc/manual/installation/installing-nspawn-container.xml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  on your system.
+ </para>
+
+ <para>
+  Create a NixOS configuration file <filename>/var/lib/machines/my-container/etc/nixos/configuration.nix</filename>.
```
Comment (markuskowa) [inline nixos/doc/manual/installation/installing-nspawn-container.xml]: I would add some explanation for the path:
```
It is important that the container root file system is under <filename>/var/lib/machines</filename>. This is the standard location where <command>machinectl</command> will look for containers. If you choose place the root into another location you need to start the container directly with <command>systemd-nspawn</command>.
```

File: nixos/doc/manual/installation/installing-nspawn-container.xml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  The file needs to have at least following options enabled:
+<programlisting>
+<xref linkend="opt-boot.isContainer"/> = true;
+<xref linkend="opt-boot.loader.initScript.enable"/> = true;
+</programlisting>
+  If your host uses <command>systemd-networkd</command> to configure the network, you can also enable <xref linkend="opt-networking.useNetworkd"/> to obtain default network configuration for your container.
```
Comment (markuskowa) [inline nixos/doc/manual/installation/installing-nspawn-container.xml]:
```suggestion
  If your host uses <command>systemd-networkd</command> to configure the network, you can also enable <xref linkend="opt-networking.useNetworkd"/> to use a default network configuration for your container.
```

**Review by markuskowa — APPROVED** [review]
The simple installer test runs locally.

**Comment by flokli** [comment]
cc @arianvp

File: nixos/tests/systemd-machinectl.nix [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  }).config.system.build.toplevel; 
+
+  containerName = "container";
+  containerRoot = "/var/lib/machines/${containerName}";
+
+in {
+  name = "systemd-machinectl";
+
+  machine = { lib, ... }: {
+    # use networkd to optain systemd network setup
```
[the remark this answers is not in the document]
Comment (ck3d) [reply nixos/tests/systemd-machinectl.nix]: thanks for reviewing, I updated the PR.

File: nixos/doc/manual/installation/installing-nspawn-container.xml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+ <para>
+  Create a NixOS configuration file <filename>/var/lib/machines/my-container/etc/nixos/configuration.nix</filename>.
+  It is important that the container root file system is under <filename>/var/lib/machines</filename>.
+  This is the standard location where <command>machinectl</command> will look for containers.
+  If you choose place the root into another location you need to start the container directly with <command>systemd-nspawn</command>.
+  The file needs to have at least following options enabled:
+<programlisting>
+<xref linkend="opt-boot.isContainer"/> = true;
+<xref linkend="opt-boot.loader.initScript.enable"/> = true;
```
Comment (arianvp) [inline nixos/doc/manual/installation/installing-nspawn-container.xml]: maybe you should note here that your should register the nix paths? What happens if you dont do this?

File: nixos/modules/installer/tools/nixos-install.sh [hunk 6]
```diff
@@ -132,8 +132,9 @@ if [[ -z $noBootLoader ]]; then
     echo "installing the boot loader..."
     # Grub needs an mtab.
     ln -sfn /proc/mounts $mountPoint/etc/mtab
-    NIXOS_INSTALL_BOOTLOADER=1 nixos-enter --root "$mountPoint" -- /run/current-system/bin/switch-to-configuration boot
+    export NIXOS_INSTALL_BOOTLOADER=1
 fi
+nixos-enter --root "$mountPoint" -- /run/current-system/bin/switch-to-configuration boot
```
Comment (arianvp) [inline nixos/modules/installer/tools/nixos-install.sh]: Why was this moved out of the `if` ? Why do we need to run `switch-to-configuration boot` ?

Comment (ck3d) [reply nixos/modules/installer/tools/nixos-install.sh]: We have to ensure that _/sbin/init_ is available, otherwise _machinectl_ do not know per default what to start.
The NixOS option _boot.loader.initScript.enable_ setups _system.build.installBootLoader_, which creates /sbin/init.
I called _switch-to-configuration_ to get the script executed.

Do you see a better way to solve this issue?

**Review by peterhoeg — APPROVED** [review]

**Review by fpletz — APPROVED** [review]

**Comment by fpletz** [comment]
Rebased onto current master and squashed some of the fixup commits. Doing a final test run now and will add a short paragraph in the release notes.

Thanks! :tada:

**Comment by fpletz** [comment]
Please do not merge this yet. There are still some issues with the functionality in this PR, probably due to the systemd bump recently.

In the generated container, some services seem to fail due to very weird reasons:

```
# systemctl -M my-container --failed                                                                                                                   ~/src/nixpkgs-tmp
  UNIT                                   LOAD   ACTIVE SUB    DESCRIPTION
● nscd.service                           loaded failed failed Name Service Cache Daemon
● systemd-journal-catalog-update.service loaded failed failed Rebuild Journal Catalog
● systemd-logind.service                 loaded failed failed Login Service
● systemd-networkd-wait-online.service   loaded failed failed Wait for Network to be Configured

LOAD   = Reflects whether the unit definition was properly loaded.
ACTIVE = The high-level unit activation state, i.e. generalization of SUB.
SUB    = The low-level unit activation state, values depend on unit type.

4 loaded units listed.
```

```
# systemctl -M my-container status systemd-logind.service                                                                                              ~/src/nixpkgs-tmp
● systemd-logind.service - Login Service
   Loaded: loaded (/nix/store/4vw3gb6dk116y38vwgjv3ymq5mfdcfli-systemd-243/example/systemd/system/systemd-logind.service; enabled; vendor preset: enabled)
  Drop-In: /nix/store/w5v9r1cl65qimpf6jggbaivgl1cgrapb-system-units/systemd-logind.service.d
           └─overrides.conf
   Active: failed (Result: exit-code) since Fri 2019-09-13 18:47:21 CEST; 11min ago
     Docs: man:systemd-logind.service(8)
           man:logind.conf(5)
           https://www.freedesktop.org/wiki/Software/systemd/logind
           https://www.freedesktop.org/wiki/Software/systemd/multiseat
  Process: 242 ExecStartPre=/sbin/modprobe -abq drm (code=exited, status=238/STATE_DIRECTORY)
  Process: 243 ExecStart=/nix/store/4vw3gb6dk116y38vwgjv3ymq5mfdcfli-systemd-243/lib/systemd/systemd-logind (code=exited, status=238/STATE_DIRECTORY)
 Main PID: 243 (code=exited, status=238/STATE_DIRECTORY)
```

Note the `/sbin/modprobe -abq drm` that failed.

Additionally, `machinectl shell` does not work for me.

**Comment by arianvp** [comment]
This seems to be similar to  as: https://github.com/systemd/systemd/issues/7605

 status=238/STATE_DIRECTORY is thrown if systemd cannot create the `StateDirectory=` directory.

**Comment by fpletz** [comment]
The `StateDirectory` was caused by me because I had to run `nixos-install` twice. I started the machine during those invocations which did the chown to the subuid/gid range but the `nixos-install` recreated some directories such as `/var/lib` as the host uid 0.

Nevertheless, `nscd` won't start because `machinectl` will default to enable user namespaces which breaks `DynamicUser` supper, see the main issue #57083 and the WIP PR #67336.

**Comment by arianvp** [comment]
@fpletz what happens if you run this as `sudo systemd-nspawn -bD /var/lib/machines/nixos` ?  If this works then I'm happy merging this as is. As the bug that seems to be stopping us is an upstream systemd issue; not something that is inherently wrong with the image

**Comment by fpletz** [comment]
@arianvp This will work (already tried it) but the attached documentation is misleading. Since we're not targeting 19.09 anymore this seems fine to me but we have to ensure we fix the documentation if this isn't fixed in systemd before 20.03.

**Comment by ck3d** [comment]
Yes I will, I prepare a PR.

File: nixos/modules/installer/tools/nixos-install.sh (hunk 6 above)
Comment (edolstra) [reply nixos/modules/installer/tools/nixos-install.sh]: Hm, why not simply *not* pass the `--no-bootloader` flag? Seems to me that you *do* want to install the boot loader (i.e. the creation of `/sbin/init`) here.

Comment (arianvp) [reply nixos/modules/installer/tools/nixos-install.sh]: because a container machinectl _directory_ will not have a `/boot` partition, but it will need an `/sbin/init` for it to be picked up by `systemd-nspawn --boot --directory <path-to-container>`

Comment (edolstra) [reply nixos/modules/installer/tools/nixos-install.sh]: `/boot` should be an implementation detail of certain boot loaders (namely GRUB and UEFI), but the `initScript` boot loader shouldn't require that.

**Comment by maxbrieiev** [comment]
So this was reverted.
Is there any way to run nixos in the container with `machinectl`/`systemd-nspawn` on a non-NixOs host?

**Comment by minijackson** [comment]
@maxbrieiev The current instructions I have are:

- Download the latest container tarball here: https://hydra.nixos.org/job/nixos/release-21.11/nixos.containerTarball.x86_64-linux

- Extract it under `/var/lib/$MACHINE_NAME`
- Then:

```bash
cp /var/lib/machines/$MACHINE_NAME/nix/store/*-etc-os-release /var/lib/machines/$MACHINE_NAME/etc`
mkdir /var/lib/machines/$MACHINE_NAME/sbin
ln -s /init /var/lib/machines/$MACHINE_NAME/sbin/init
chattr -i /var/lib/machines/$MACHINE_NAME/var/empty
```
